src/primitives/box.jl

```diff
@@ -33,6 +33,23 @@ measure(b::Box) = prod(b.max - b.min)
 diagonal(b::Box) = norm(b.max - b.min)
 sides(b::Box) = b.max - b.min
 
+"""
+Return all corners of the box.
+"""
+@generated function vertices(b::Box)
+  list = sort(collect(distinct(IterTools.subsets(repeat(1:2, paramdim(b)), Val{paramdim(b)}()))))
+  P = Point{embeddim(b),coordtype(b)}
+  ex = Expr(:ref, P)
+  map(list) do idxs
+    next_tuple = Expr(:tuple, (:(coords[$(idxs[i])][$i]) for i in eachindex(idxs))...)
+    push!(ex.args, next_tuple)
+  end
```

Comment:
So you are looping over all possible indices right? Can´t we use CartesianIndices for that in a cleaner way?

```julia
julia> for I in CartesianIndices((2,2))
         @show I
       end
I = CartesianIndex(1, 1)
I = CartesianIndex(2, 1)
I = CartesianIndex(1, 2)
I = CartesianIndex(2, 2)
```

I think it already uses @generated behind the scenes.

Reply:
Oooh I didn't know that. Indeed that simplifies everything :sweat_smile:

Reply:
Ah maybe not exactly, in 3D I need to generate
```julia
[[1, 1, 2], [1, 2, 1], [2, 1, 1], [2, 1, 2], [2, 2, 1], [2, 2, 2]]
```

Reply:
Also we need to make sure that the CCW orientation is respected in 2D I guess? So that we can feed a Box into a mesh algorithm for example. Perhaps this level of automation needs to be more manual than what we would like it to be. Maybe we should cover the 2D and 3D cases manually for safety?

Reply:
Got it working. That's awesome, thanks for suggesting that.

Reply:
I'll look into that.

Reply:
In 2D it should work out well with the ordering given by `CartesianIndices`. In 3D not so much if you want triangles to be CCW and recover all the cube. We need an ordering scheme to link the vertices in that case.

Reply:
To be more accurate, in 2D it works well only if you consider triangle strips (like [shown here](https://www.khronos.org/registry/vulkan/specs/1.2-khr-extensions/html/chap20.html#drawing-triangle-strips))

Reply:
For the sake of clarity, should we use simple code for 2D and 3D for now? We are still working out this CCW details and I have no idea yet how they will play in 3D. So if we could have simple code that is easy to read we could think about generalizations in the future. It is a cool code but won´t necessarily attract many contributors :)

Reply:
I am not sure what you want, do you mean writing what the generated function does by hand for 2D/3D cases or you're talking about triangulating the box?

Reply:
If you're talking about the generated function, then we can put the expression processing code inside another function and expand it like [there](https://stackoverflow.com/questions/66402105/any-way-to-expand-a-generated-function-in-julia) to be able to get the generated expression.

Reply:
I mean we could keep it simple, I don´t see a real need for generated functions and more advanced-looking code when we can't deal with higher dimensions yet. Keeping it simple will help new contributors to suggest improvements. Let's keep it simple for now. That way we reduce the need for external libraries such as IterTools which is currently only used for `ivec`.

Reply:
I don't think it would be useful. Generated functions are very good when you know the logic up-front and don't really want to type in all the code yourself. That way you don't need to look at two different functions for 2D and 3D, particularly when `Box` is already defined very generically. And we support 4D if someone decides he needs to go in 4D (unlikely, but possible); writing 3 functions (2D, 3D by hand and keeping this generated function for higher-D) would be a mess to maintain, and not keeping the generated function would lead to a loss of (admittedly minor) features.

Also, it does quite a simple thing and gives natural results, and I don't really see why we'd want to change it in the future. And if someone decides to change it, a manual version may not be easier since you would need to reconstruct this logic in your head anyway. This `CartesianIndices` logic is very central and if one knows it upfront it will be easier to see how we loop over indices.

Lastly we don't use IterTools at all there (`CartesianIndices` removed the dependency).

Reply:
To mitigate the difficulty that one may have to understand this function, we can add an example in the docstring which shows what it does in 2D and 3D.